Testnet/Airdrop/Airdrop/Airdrop.cs

```diff
@@ -0,0 +1,202 @@
+        ulong totalSupply,
+        ulong endBlock
+    ) : base(smartContractState)
+    {
+        TotalSupply = totalSupply;
+        TokenContractAddress = tokenContractAddress;
+        EndBlock = endBlock;
+        Owner = Message.Sender;
+    }
+
+    /// <summary>The contract address of the token that will be distributed. This smart contracts address must hold the TotalSupply at this address.</summary>
+    public Address TokenContractAddress
+    {
+        get => PersistentState.GetAddress(nameof(TokenContractAddress));
+        private set => PersistentState.SetAddress(nameof(TokenContractAddress), value);
+    }
+
+    /// <summary>The total supply of the token that will be distributed during this airdrop.</summary>
+    public ulong TotalSupply
+    {
+        get => PersistentState.GetUInt64(nameof(TotalSupply));
+        private set => PersistentState.SetUInt64(nameof(TotalSupply), value);
+    }
+
+    /// <summary>
+    /// Used to automatically close registration based on a specified block number. If the EndBlock is 0, the
+    /// registration period for the airdrop must be manually ended by the owner using <see cref="CloseRegistration"/>
+    /// </summary>
+    public ulong EndBlock
+    {
+        get => PersistentState.GetUInt64(nameof(EndBlock));
+        private set => PersistentState.SetUInt64(nameof(EndBlock), value);
+    }
+
+    /// <summary>Address of the owner of this contract, used to authenticate some methods.</summary>
+    public Address Owner
+    {
+        get => PersistentState.GetAddress(nameof(Owner));
+        private set => PersistentState.SetAddress(nameof(Owner), value);
+    }
+
+    /// <summary>The total number of registrants for this airdrop.</summary>
+    public ulong NumberOfRegistrants
+    {
+        get => PersistentState.GetUInt64(nameof(NumberOfRegistrants));
+        private set => PersistentState.SetUInt64(nameof(NumberOfRegistrants), value);
+    }
+
+    /// <summary>Calculates and sets the amount to distribute to each registrant.</summary>
+    public ulong AmountToDistribute
+    {
+        get
+        {
+            ulong amount = PersistentState.GetUInt64(nameof(AmountToDistribute));
+            if (amount == 0 && RegistrationIsClosed)
+            {
+                amount = TotalSupply / NumberOfRegistrants;
+                AmountToDistribute = amount;
+            }
+
+            return amount;
+        }
+
+        private set => PersistentState.SetUInt64(nameof(AmountToDistribute), value);
+    }
+
+    /// <summary>Returns whether or not the registration period is closed.</summary>
+    public bool RegistrationIsClosed
+    {
+        get
+        {
+            bool isClosed = PersistentState.GetBool(nameof(RegistrationIsClosed));
+            if (!isClosed && EndBlock > 0)
+            {
+                isClosed = Block.Number > EndBlock;
+                RegistrationIsClosed = isClosed;
+            }
+
+            return isClosed;
```

Comment:
Would recommend moving this logic out of the getter. The main reason is that it makes the contract harder to reason about. There too many things are happening here - checking the block number, checking whether the registration is closed, and updating the `RegistrationIsClosed` property.

Consider breaking it into a few discrete steps:
- `RegistrationIsClosed` as a basic get/set property.
- Don't update `RegistrationIsClosed` when the block number is exceeded. Instead, compute this with a second property, something like `public bool CanRegister => RegistrationIsClosed || Block.Number > EndBlock;`.